src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/ElectricityMapsClient.cs

```diff
@@ -0,0 +1,210 @@
+        };
+
+        return await GetCurrentForecastAsync(parameters);
+    }
+
+    // Internal call to check for allowed zones and then make GET request to Forecast endpoint
+    private async Task<ForecastedCarbonIntensityData> GetCurrentForecastAsync(Dictionary<string, string> parameters)
+    {
+        await CheckZonesAllowedForPath(Paths.Forecast, parameters);
+        using (var result = await this.MakeRequestGetStreamAsync(Paths.Forecast, parameters))
+        {
+            return await JsonSerializer.DeserializeAsync<ForecastedCarbonIntensityData>(result, _options) ?? throw new ElectricityMapsClientException($"Error getting forecasted data");
+        }
+    }
+
+    private async Task<HttpResponseMessage> GetResponseAsync(string uriPath)
+    {
+        var response = await _client.GetAsync(uriPath, HttpCompletionOption.ResponseHeadersRead);
+        if (!response.IsSuccessStatusCode)
+        {
+            var content = await response.Content.ReadAsStringAsync();
+            _log.LogError("Error getting data from ElectricityMaps {uriPath}. StatusCode: {statusCode}. Content: {content}", uriPath, response.StatusCode, content);
+            throw new ElectricityMapsClientHttpException($"Error requesting {uriPath} - Content: {content}", response);
+        }
+        return response;
+    }
+
+    private async Task<Stream> MakeRequestGetStreamAsync(string path, Dictionary<string, string>? parameters = null)
+    {
+        var url = path;
+        if (parameters is not null)
+        {
+            url = BuildUrlWithQueryString(path, parameters);
+        }
+        _log.LogInformation("Requesting data using url {url}", url);
+        var response = await this.GetResponseAsync(url);
+        return await response.Content.ReadAsStreamAsync();
+    }
+
+    private string BuildUrlWithQueryString(string url, IDictionary<string, string> queryStringParams)
+    {
+        if (_log.IsEnabled(LogLevel.Debug))
+        {
+            _log.LogDebug("Attempting to build a url using url {url} and query string parameters {parameters}", url, string.Join(";", queryStringParams.Select(k => $"\"{k.Key}\":\"{k.Value}\"")));
+        }
+        var query = HttpUtility.ParseQueryString(string.Empty);
+        foreach(var kvp in queryStringParams)
+        {
+            query[kvp.Key] = kvp.Value;
+        }
+        var result = $"{url}?{query}";
+        if (_log.IsEnabled(LogLevel.Debug))
+        {
+            _log.LogDebug("Built url {result} from url {url} and query string parameters {parameters}", result, url, string.Join(";", queryStringParams.Select(k => $"\"{k.Key}\":\"{k.Value}\"")));
+        }
+        return result;
+    }
+
+    // Checks the current supported client's endpoint paths.
+    private async Task CheckZonesAllowedForPath(string path, Dictionary<string, string> parameters)
+    {
+        // Trial user won't have access to zones endpoint, exit
+        if (this._configuration.BaseUrl.Contains(BaseUrls.TrialBaseUrl)) return; 
```

Comment:
The goal is to avoid the performance hit of trying to access ElectricityMaps when we know for a fact it will fail. Instead, just exit the check all together, as the user doesn't need this check, it's more to give them better error messages if the subsequent call fails

Reply:
@JenMadiedo will workaround the 500 failure post convo with bill